### ISSUE TITLE
options(languageserver.formatting_style) doesn't work in VSCode

### ISSUE THREAD

**Issue by ColinFay** [REditorSupport/languageserver#503]
Hey, 

Wasn't sure if this needed to be opened in https://github.com/REditorSupport/vscode-R or here, feel free to move the issue. 

I've followed the recommendation for setting custom styler options. 

As a reprex, I've set my .Rprofile as 

```r
options(languageserver.formatting_style = function(options) {
    style <- styler::tidyverse_style(indent_by = options$tabSize)
    style$token$force_assignment_op <- NULL
    style
})
```

Then when I saved a file, I still get the `=` replaced by `<-`

Before save

![Screenshot 2021-12-03 at 11 49 25](https://user-images.githubusercontent.com/17936236/144590539-4dd2ceb0-1238-4a46-b92e-5432fe75b608.png)

After save

![Screenshot 2021-12-03 at 11 49 29](https://user-images.githubusercontent.com/17936236/144590561-03477ce4-3af7-45c8-9bec-1b5172176bb5.png)

Am I missing something?

Thanks for your help, 
Colin

**Comment by renkun-ken** [comment]
I cannot reproduce it.

Did you reload window in VS Code to restart languageserver to make the formatting style take effect?

**Comment by ColinFay** [comment]
I tried setting the option manually but it seems that the formatter function is not called ?

```r
options(languageserver.formatting_style = function(options) {
    cli::cat_rule("In formatting_style")
    style <- styler::tidyverse_style(indent_by = options$tabSize)
    style$token$force_assignment_op <- NULL
    style
})
```

![Screenshot 2021-12-03 at 13 43 52](https://user-images.githubusercontent.com/17936236/144604616-94c66aba-ec85-4e96-a300-db6bbbbb23a8.png)

**Comment by renkun-ken** [comment]
I notice that in your workspace folder, there is a `.Rprofile`. Then language server will use this file rather than `~/.Rprofile` as it is started from the workspace folder.

Executing the code does not work because it only sends code to the user R terminal, but language server is a background process which loads the `.Rprofile` on startup.

**Comment by renkun-ken** [comment]
In your output pane, the R language server log shows that two language servers are started. It is because you are editing external files (probably `~/.Rprofile`) rather than the workspace `.Rprofile`. You should edit the workspace `.Rprofile` or remove it.

**Comment by ColinFay** [comment]
I'm editing / saving the workspace `.Rprofile`. 

The one displayed on the screenshot is the project-specific `.Rprofile`, not the global one.

![Screenshot 2021-12-03 at 14 16 18](https://user-images.githubusercontent.com/17936236/144608731-8d632ad3-f931-4e42-893d-74a39677e109.png)

**Comment by eitsupi** [comment]
Is it possible that the last line of the Rprofile is not functioning because there is no newline?
https://stackoverflow.com/questions/18469416/r-seems-to-ignore-the-last-line-of-rprofile

**Comment by ColinFay** [comment]
Thanks @eitsupi, I tried that but that didn't change anything :(

**Comment by renkun-ken** [comment]
Would you like to enable `r.lsp.debug` and post the log of R Language Server. It has some startup info.

**Comment by renkun-ken** [comment]
You could put the following in your `.Rprofile`

```r
cat("[start]", Sys.getpid(), paste0(commandArgs(), collapse = " "), file = "~/rprofile.log", append = TRUE)
options(languageserver.formatting_style = function(options) {
    style <- styler::tidyverse_style(indent_by = options$tabSize)
    style$token$force_assignment_op <- NULL
    style
})
cat("[end]", "\n", file = "~/rprofile.log", append = TRUE)
```

and restart vscode. And then you could put the content of `~/rprofile.log` here and see if the `.Rprofile` is acutally sourced by the languageserver process.

**Comment by ColinFay** [comment]
Hey, 

Here is the output: 

```
colin:test colin$ cat ~/rprofile.log
[start] 18394 /Library/Frameworks/R.framework/Resources/bin/exec/R --no-save --no-restore[end] 
```

And the output with r.lsp.debug turned on.

```
R Language Server (18135) started
[2021-12-07 10:34:45.072] received:  Content-Length: 263

[2021-12-07 10:34:45.101] received payload.

[2021-12-07 10:34:45.101] handling request:  textDocument/codeAction

[2021-12-07 10:34:45.101] text_document_code_action:  {
  "textDocument": {
    "uri": "file:///Users/colin/Seafile/documents_colin/R/trash/test/test.R"
  },
  "range": {
    "start": {
      "line": 0,
      "character": 0
    },
    "end": {
      "line": 0,
      "character": 0
    }
  },
  "context": {
    "diagnostics": []
  }
}

[2021-12-07 10:34:45.107] document_code_action_reply:  {
  "uri": "file:///Users/colin/Seafile/documents_colin/R/trash/test/test.R",
  "range": {
    "start": {
      "row": 0,
      "col": 0
    },
    "end": {
      "row": 0,
      "col": 0
    }
  },
  "context": {
    "diagnostics": []
  },
  "result": []
}

[2021-12-07 10:34:45.111] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.113] received:  Content-Length: 163

[2021-12-07 10:34:45.123] received payload.

[2021-12-07 10:34:45.124] handling request:  textDocument/documentColor

[2021-12-07 10:34:45.126] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.129] received:  Content-Length: 162

[2021-12-07 10:34:45.140] received payload.
[2021-12-07 10:34:45.140] handling request:  textDocument/documentLink
[2021-12-07 10:34:45.141] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.143] received:  Content-Length: 162

[2021-12-07 10:34:45.168] received payload.

[2021-12-07 10:34:45.169] handling request:  textDocument/foldingRange

[2021-12-07 10:34:45.172] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.276] received:  Content-Length: 263

[2021-12-07 10:34:45.289] received payload.

[2021-12-07 10:34:45.292] handling request:  textDocument/codeAction

[2021-12-07 10:34:45.293] text_document_code_action:  {
  "textDocument": {
    "uri": "file:///Users/colin/Seafile/documents_colin/R/trash/test/test.R"
  },
  "range": {
    "start": {
      "line": 0,
      "character": 5
    },
    "end": {
      "line": 0,
      "character": 5
    }
  },
  "context": {
    "diagnostics": []
  }
}

[2021-12-07 10:34:45.296] document_code_action_reply:  {
  "uri": "file:///Users/colin/Seafile/documents_colin/R/trash/test/test.R",
  "range": {
    "start": {
      "row": 0,
      "col": 5
    },
    "end": {
      "row": 0,
      "col": 5
    }
  },
  "context": {
    "diagnostics": []
  },
  "result": []
}

[2021-12-07 10:34:45.300] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.614] received:  Content-Length: 162

[2021-12-07 10:34:45.626] received payload.

[2021-12-07 10:34:45.627] handling request:  textDocument/documentLink

[2021-12-07 10:34:45.629] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.633] received:  Content-Length: 163

[2021-12-07 10:34:45.644] received payload.

[2021-12-07 10:34:45.644] handling request:  textDocument/documentColor

[2021-12-07 10:34:45.648] deliver:  ["Response", "Message", "R6"]

[2021-12-07 10:34:45.863] received:  Content-Length: 162

[2021-12-07 10:34:45.876] received payload.

[2021-12-07 10:34:45.877] handling request:  textDocument/foldingRange

[2021-12-07 10:34:45.879] deliver:  ["Response", "Message", "R6"]
```

**Comment by ColinFay** [comment]
That's probably related, but, it seems that the auto-formatting is done even if I don't launch an R session in the terminal

**Comment by renkun-ken** [comment]
The language server process does not appear in your `~/rprofile.log`, which indicates that the language server does not load the `~/.Rprofile` at all.

**Comment by ColinFay** [comment]
@renkun-ken Argh. I'm not sure what to do to debug that ?

**Comment by ColinFay** [comment]
I did the following : 

+ put the code in both Rprofile

+ Closed all VSCode instances 

+ Reopen

+ Clicked on the "Attach R" button in VSCode 

![Screenshot 2021-12-07 at 17 00 17](https://user-images.githubusercontent.com/17936236/145063228-b42b18bd-4663-4956-9203-0a2658ca4a23.png)

Here is the content of `~/rprofile.log`

```
colin:test colin$ cat ~/rprofile.log
<project> [start] 30375 /Library/Frameworks/R.framework/Resources/bin/exec/R --no-save --no-restore[end]
```

**Comment by renkun-ken** [comment]
Did you open an R file and activate the language server?

**Comment by ColinFay** [comment]
No, I launched the language server without opening any file here. 

The output in the rprofile log is the same even if I open a script.

**Comment by renkun-ken** [comment]
Maybe I'm not saying it clearly.

The language server is started by vscode-R when you open any R file. If you don't open any file, then no language server is started, and we can't see the log in `~/rprofile.log` and know which `.Rprofile` is acutally used by the language server process.

If you open a R file and the language server process does not show in `~/rprofile.log`, then the language server process is not using any of the `.Rprofile` we know. Did you alter the `r.lsp.args` settings to include something like `--vanilla` which prevents it from loading `.Rprofile`?

**Comment by ColinFay** [comment]
Hey, 

Sorry to get back to this so late. 

I started on a fresh laptop and still have this issue: the formatting is the default one and what is passed to the `option` is not launched, so I decided to dig deeper. 

I managed to get this to work by clearing all the cache in styler : 

```r
styler::cache_clear()
```

Modify my global .Rprofile.

Close everything

Restart VSCode, 

and tadaa 🎉

Thanks @renkun-ken for taking the time to help. 

Cheers, 
Colin
